### PR TITLE
Use Primitive#upload instead of copy/paste its code

### PR DESCRIPTION
**Pull request by DmitrySharabin**
I just want to say that it's OK if we won't merge this PR since it has to deal with some aspects of properties, like an attribute and a property value (see the comments in the code).

If we decide to leave the previously merged version, it might be a good idea to use line 121 from this PR and send a separate PR with it. What do you think?

### CODE REVIEW

**Comment by DmitrySharabin** [comment]
@LeaVerou Hey Lea! 👋🏻
Will you please have a look at this PR? I have an open Gestalt because ot it. :)

File: markdown/mavo-markdown.js [hunk 1]
```diff
@@ -110,46 +110,46 @@ Mavo.Elements.register("markdown", {
...
+						env.editor.value = oldValue;
 
 						// Enable editor and replace the placeholder with the image element
 						env.editor.disabled = false;
 
-						const t = evt.target;
 						t.focus();
-						t.selectionStart = t.selectionEnd - placeholder.length;
-						t.setSelectionRange(t.selectionStart, t.selectionEnd);
+						t.setSelectionRange(cursorPosition, cursorPosition + placeholder.length);
```
Comment (LeaVerou) [inline markdown/mavo-markdown.js:152]: What about when there's a selection?
What about after the upload?

File: markdown/mavo-markdown.js [hunk 2]
```diff
@@ -110,46 +110,46 @@ Mavo.Elements.register("markdown", {
 
 		// Image upload
 		env.editor.addEventListener("paste", async evt => {
-			if (env.context.mavo.uploadBackend && self.FileReader) {
+			if (this.mavo.uploadBackend && self.FileReader) {
 				// Look for the first image in the clipboard
 				const item = Array.from(evt.clipboardData.items).find(item => item.kind == "file" && item.type.indexOf("image/") === 0);
```
Comment (LeaVerou) [inline markdown/mavo-markdown.js:115]: This also seems like it should live in one place and not be repeated everywhere uploads via paste are allowed, but that's minor.

File: markdown/mavo-markdown.js [hunk 3]
```diff
@@ -110,46 +110,46 @@ Mavo.Elements.register("markdown", {
...
 					if (name === "") {
 						name = defaultName;
 					}
 
 					if (name !== null) {
+						const t = evt.target;
+
 						// Disable editor and show placeholder while waiting for an image to upload
-						const placeholder = `![${env.context.mavo._("uploading")}...]()`;
+						const placeholder = `![${this.mavo._("uploading")}...]()`;
```
Comment (LeaVerou) [inline markdown/mavo-markdown.js:132]: What happens if a user tries to save while an image is uploading?
Github disables the submit button until the upload finishes. Do we want to do something similar? Do we want to perhaps introduce a confirmation?

**Review by LeaVerou — CHANGES_REQUESTED** [review]
I added some comments, though since this is a clear improvement over the previous code I'm tempted to merge as-is. Up to you!

**Comment by LeaVerou** [comment]
Also, do note that simply replacing the entire text drops your undo stack. There are workarounds, but not for all browsers. 

See

https://stackoverflow.com/questions/19814465/is-it-possible-to-insert-text-in-textarea-and-update-undo-redo-queuehttps://stackoverflow.com/questions/44471699/how-to-make-undo-work-in-an-html-textarea-after-setting-the-value

and many others

**Comment by DmitrySharabin** [comment]
Thank you, Lea, for all your comments. I learn a lot from you. I'll work through them and will make some improvements.

**Comment by DmitrySharabin** [comment]
As I mentioned in #76, I'm going to close this PR with a pure heart. I learned a lot from it. :)

**Pull request closed without merging**